import sys
from Bio import SeqIO
from Bio.Blast import Record



clusterFile = sys.argv[1]
count = 0
record_iterator = SeqIO.parse("Cluster_%s" %count, "fasta")
for i in count:
    for seq_record in SeqIO.parse("Cluster_%s" %i, "fasta"):
        if seq_record.id == 0
            print(seq_record.id)
            print(repr(seq_record.seq))
            print(len(seq_record))




record_iterator = SeqIO.parse("Cluster_%s" %count, "fasta")
first_record = next(record_iterator)
print(first_record.id)
print(first_record.description)

second_record = next(record_iterator)
print(second_record.id)
print(second_record.description)

record_iterator = SeqIO.parse("Cluster_%s" %count, "fasta")

first_record = next(record_iterator)
print(first_record.id)
print(first_record.description)

second_record = next(record_iterator)
print(second_record.id)
print(second_record.description)















result_handle = open(clusterFile)
listOfRecords = Record.parse(result_handle)




for gene in listOfRecords:
	for i in gene.alignments)):
        if     
        
        
        output = "> %s\n%s\n" % (title, sequence.replace('-', ''))







with open(("Cluster_%s" %count), 'r') as cluster:
    readline()
    if
    for eachgene in cluster
        gene= 
    dictprot05 = SeqIO.index("05.fa.txt.pfa", "fasta")

count = 0
for count in 
    
